File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head = None
        cur = None
        while ((list1 is not None) or (list2 is not None)):
            # find min
            min = None
            if list1 is None:
                min = list2
                list2 = list2.next
            elif list2 is None:
                min = list1
                list1 = list1.next
            elif list1.val <= list2.val:
                min = list1
                list1 = list1.next
            else:
                min = list2
                list2 = list2.next

            print(min.val)
            # Add to list
            if head is None:
                head = min
                cur = head
            else:
                cur.next = min
                cur = cur.next
            
        return head
            

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        // System.out.println(lists[0].val);
        ListNode newHead = null;
        ListNode tail = null;
        /*
        while (each linkedlist cur is not null) {
            int min = min(ith linkedlist head)
            ith linkedlist head = ith linkedlist.next
        }
        */

        while (condition(lists)) {
            int min = 1001;
            int index = -1;
            for (int i = 0; i < lists.length; i++) {
                if (lists[i] != null) {
                    if (lists[i].val < min) {
                        min = lists[i].val;
                        index = i;
                    }
                }
            }

            ListNode node = new ListNode(lists[index].val);
            if (newHead == null || tail == null) {
                newHead = node;
                tail = node;
            } else {
                tail.next = node;
                tail = node;
            }
            lists[index] = lists[index].next;
        }
        return newHead;
    }

    private boolean condition(ListNode[] lists) {
        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null) {
                return true;
            }
        }
        return false;
    }
}
